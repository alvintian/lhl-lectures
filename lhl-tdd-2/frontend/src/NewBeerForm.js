import React, { Component } from "react";

class NewBeerForm extends Component {
  constructor(props) {
    super(props);
    this.state = { name: "", style: "", abv: "" };
  }

  handleChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };

  handleSubmit = event => {
    event.preventDefault();
    this.props.onSubmit(this.state);
  };

  render() {
    const { name, style, abv } = this.state;
    return (
      <form onSubmit={this.handleSubmit} noValidate>
        <div>
          <label labelFor="name">Name</label>
          <input
            type="text"
            name="name"
            id="name"
            value={name}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <label labelFor="style">Style</label>
          <input
            type="text"
            name="style"
            id="style"
            value={style}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <label labelFor="abv">ABV</label>
          <input
            type="text"
            name="abv"
            id="abv"
            value={abv}
            onChange={this.handleChange}
          />
        </div>
        <button type="submit">Create Beer</button>
      </form>
    );
  }
}

export default NewBeerForm;
