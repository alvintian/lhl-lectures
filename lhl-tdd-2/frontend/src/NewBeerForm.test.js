import React from "react";
import { shallow } from "enzyme";

import "./testSetup";
import NewBeerForm from "./NewBeerForm";

describe("<NewBeerForm/>", () => {
  describe(".render()", () => {
    it("renders a new form", () => {
      const wrapper = shallow(<NewBeerForm />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe(".handleUpdate(event)", () => {
    it("updates the state based on the name input value", () => {
      const wrapper = shallow(<NewBeerForm />);
      expect(wrapper.state()).toEqual({ name: "", abv: "", style: "" });
      wrapper
        .find("#name")
        .simulate("change", { target: { name: "name", value: "Bud" } });
      wrapper
        .find("#style")
        .simulate("change", { target: { name: "style", value: "Lager" } });
      wrapper
        .find("#abv")
        .simulate("change", { target: { name: "abv", value: "4.5" } });
      expect(wrapper.state()).toEqual({
        name: "Bud",
        abv: "4.5",
        style: "Lager"
      });
    });
  });

  describe(".onSubmit(data)", () => {
    const onSubmitMock = jest.fn();
    const wrapper = shallow(<NewBeerForm onSubmit={onSubmitMock} />);
    const state = { name: "Bud", style: "Lager", abv: "4.0" };
    const eventMock = { preventDefault: jest.fn() };

    beforeAll(() => {
      wrapper.setState(state);
      wrapper.find("form").simulate("submit", eventMock);
    });

    it("prevents the default action from occuring", () => {
      expect(onSubmitMock).toHaveBeenCalledWith(state);
    });

    it("on form submit it returns the data to the onSubmit() callback", () => {
      expect(eventMock.preventDefault).toHaveBeenCalled();
    });
  });
});
