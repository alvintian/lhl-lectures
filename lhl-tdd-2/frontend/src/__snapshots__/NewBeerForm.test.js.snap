// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewBeerForm/> .render() renders a new form 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div>
    <label
      labelFor="name"
    >
      Name
    </label>
    <input
      id="name"
      name="name"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      labelFor="style"
    >
      Style
    </label>
    <input
      id="style"
      name="style"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      labelFor="abv"
    >
      ABV
    </label>
    <input
      id="abv"
      name="abv"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <button
    type="submit"
  >
    Create Beer
  </button>
</form>
`;
